### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by jonsberndt** [JSBSim-Team/jsbsim#366] [Ideas]
I've wondered if there might be a way to implement a particular testing capability within JSBSim. I'm thinking about unit testing ... sort of. Here's an example: say I want to output the entire set of aerodynamic characteristics for a vehicle. Say that the aero coefficients are functions of alpha, beta, and qbar (as well as characteristic lengths, etc.). I'd like to be able to independently test a coefficient and make sure the output is as expected. However, we can't unit test an individual piece of code, nor easily load the XML data into an individual coefficient for test. The code has to be tested as it is, integrated as a whole with all the other code. 

Alpha, beta, and qbar are all calculated at runtime and cannot themselves be overridden in order to drive the aerodynamic coefficients based on controlled inputs. That got me wondering if within each model (or any piece of code) if there could/should be a test() function that is unique to that code and permits the execution of the code in a controlled way, with specified inputs, and output as needed for test. That would probably be a major feature. 

Is there another way? Here's another example: I'd like to be able to run the Inertial code and get out the gravity for a specific location and altitude. I'd like to just run the code for a set of input conditions, running only the gravity model, and verify outputs. 

A testing feature could also be useful for propulsion systems. And so on.

Some ideas to consider if this was to be pursued: the existing code should not be made harder to read, and any testing code added should be relatively lightweight and unobtrusive.

I've also wondered if the goals could be reached solely through the use of scripting and XML. For instance, for an aerodynamic coefficient, perhaps the coefficient could be designed in XML to be a function not of alpha, but of alpha_prime, and similarly for beta and qbar. And at the beginning of the aerodynamics section there could be functions that defined for these *_prime parameters which would take into account whether a "testing" property is set true or not. If "testing" is set to false (perhaps by setting that property value to false on the command line for JSBSim standalone) then alpha_prime would take the value of alpha, otherwise it would take the value as set via scripting or command line. The general idea behind this is to permit setting the dependent values for aero coefficients -OR- letting the aero coefficients be calculated normally, depending on whether one is testing or not.

I haven't fleshed these ideas out totally, but wanted to throw them out there for comment. Perhaps someone has already addressed this.

**Comment by bcoconni** [comment]
FYI we already have C++ unit tests in `tests/unit_tests`. At the moment we are only testing `FGColumnVector3`, `FGGroundCallback`, `FGInertial`, `FGInitialConditions` (partially), `FGJSBBase`, `FGLocation`, `FGMatrix33`, and `FGQuaternion`.

We also have integration tests in `tests`. These are written in Python because the language has all sorts of libraries (network, numerical analysis, etc.) but they could be written in C++ if one would like to.

All of these tests are automatically run and checked at each commit and pull request to make sure we do not have any regression (see for example the [checks ran after Sean's last pull request](https://github.com/JSBSim-Team/jsbsim/actions/runs/449152634))

If you plan to test JSBSim correctness, I would rather prefer that we continue to extend the current tests (unit tests and integration tests) rather than starting a new means of testing from the ground up.
> Is there another way? Here's another example: I'd like to be able to run the Inertial code and get out the gravity for a specific location and altitude. I'd like to just run the code for a set of input conditions, running only the gravity model, and verify outputs.

This could be easily be done using Python scripting with a few modifications to [`python/jsbsim.pyx.in`](https://github.com/JSBSim-Team/jsbsim/blob/master/python/jsbsim.pyx.in). Something like:
```python
import jsbsim

fdm = jsbsim.FGFDMExec()
inertial = fdm.GetInertial()
```
> I've also wondered if the goals could be reached solely through the use of scripting and XML.

I guess this could be interesting for users who want to test their model without having to resort to Python scripting ?

**Reply by jonsberndt** [reply]
The idea is to not so much test the code, but to test an individual model. That requires the aircraft and other configuration files to be read in first. Then somehow output data that shows that under all expected conditions that the model is correct. So, for the aero model, for instance, one would want to vary the independent variables (normally alpha, beta, qbar, etc.) across their range. 

I suspect this could be done in a number of ways, but ... what is the best way?

**Comment by seanmcleod** [comment]
Scripting based test setup I tried out based on @jonsberndt's suggestion regarding 'prime' inputs above.

Had to make some changes/additions to the sandbox testing functionality in `JSBSim_utils`.

But pretty simple, basically copy the aircraft config file into a sandbox and then perform a simple text based search and replace to rename the input properties that need to be used to test the aerodynamic model you want to test.

```python
import os
from JSBSim_utils import CreateFDM2, CopyAircraftConfig, SandBox

testInputs = ['aero/qbar-psf', 'aero/alpha-rad']

aircraft = '737'

# Setup sandbox
sandbox = SandBox()
currentdir = os.getcwd()
os.chdir(sandbox())

# Copy aircraft config file from JSBSim into sandbox
aircraftConfigFilePath = CopyAircraftConfig(aircraft, sandbox)

# Create FDM referencing the sandbox
fdm = CreateFDM2(sandbox)

# Create test input properties
for input in testInputs:
    fdm['test/' + input] = 0.0

# Replace inputs in aircraft config with test versions
file = open(aircraftConfigFilePath, 'r+')
contents = file.read()
contents = contents.replace('Dave Culp', 'Chuck Yeager')
for input in testInputs:
    contents = contents.replace(input, 'test/' + input)
file.seek(0)
file.write(contents)
file.truncate()
file.close()

# Load up modified aircraft config
fdm.load_model(aircraft)

# Perform tests
testModel = 'aero/coefficient/CLalpha'
# (qbar-psf, alpha-rad, lift)
groundTruthData = [ (100, 0.1, 89422.65), (100, 0.2, 150671.04), (200, 0.1, 178845.30), (200, 0.2, 301342.08) ]

for testData in groundTruthData:
    for index, testInput in enumerate(testInputs):
        fdm['test/' + testInput] = testData[index]
    value = fdm[testModel]
    if abs(value - testData[-1]) > 0.1:
        print('Error model mismatch {0} : {1}'.format(testData, value))

# Cleanup
del fdm
fdm = None
os.chdir(currentdir)
sandbox.erase()
```

So no need to edit the aircraft config file and no need to make any source code changes to JSBSim. However this won't work for all models, but should work for most/all aerodynamic models that are simply functions based on property inputs that you want to vary.

**Reply by jonsberndt** [reply]
This is fascinating. First of all, I hadn't realized that JSBSim had been "python-ized" to this degree. I think that's fantastic. I'm not very familiar with python, though I am using Julia in simulation development at the moment. 

I did implement a testing capability with xml for a particular model and that worked well. Though, I think this python approach is cleaner. I agree with both you and Bertrand that there is probably no need for code changes. 

Is this python capability documented somewhere? Like ... what is a sandbox?

Thanks.

**Reply by bcoconni** [reply]
> Like ... what is a sandbox?

A sandbox is basically an object that creates a temporary folder that will contain all the output of JSBSim. Once the test is completed, the sandbox is destroyed and the output files deleted. The purpose is to avoid the `tests` folder to be filled with loads of CSV and XML files.

**Comment by seanmcleod** [comment]
Don't think there is much documentation other than the source code for things like the sandbox etc.

But I'd suggest taking a look at a couple of the python based tests we have in the `tests` sub-directory to get a feel for the sort of python support and the sorts of things we've been testing.

For example after we made some fairly big changes to the standard atmosphere model - 

https://github.com/JSBSim-Team/jsbsim/blob/master/tests/TestStdAtmosphere.py

After we added support for stability axes etc. - 

https://github.com/JSBSim-Team/jsbsim/blob/master/tests/TestAeroFuncFrame.py

Testing the actuator model after a bug report - 

https://github.com/JSBSim-Team/jsbsim/blob/master/tests/TestActuator.py

**Comment by bcoconni** [comment]
@jonsberndt 
> [...] though I am using Julia in simulation development at the moment.

I was telling myself may be we could provide an interface of JSBSim with Julia just like we do for Python ? I am not familiar myself with Julia but I'm pretty sure we could get some new users if we provide support to Julia !

I have added the language to the list proposed in issue #70
